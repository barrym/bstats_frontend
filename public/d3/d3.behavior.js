(function(){d3.behavior = {};
// TODO unbind zoom behavior?
// TODO unbind listener?
d3.behavior.zoom = function() {
  var x = 0,
      y = 0,
      z = 0,
      panning,
      zooming,
      locations = {}, // identifier -> location
      last = 0,
      target,
      targuments,
      event = d3.dispatch("zoom");

  // mousewheel events are totally broken!
  // https://bugs.webkit.org/show_bug.cgi?id=40441
  // not only that, but Chrome and Safari differ in re. to acceleration!
  var div = d3.select("body").append("div")
      .style("visibility", "hidden")
      .style("top", 0)
      .style("height", 0)
      .style("width", 0)
      .style("overflow-y", "scroll")
    .append("div")
      .style("height", "2000px")
    .node().parentNode;

  function zoom() {
    this
        .on("mousedown.zoom", mousedown)
        .on("mousewheel.zoom", mousewheel)
        .on("DOMMouseScroll.zoom", dblclick)
        .on("dblclick.zoom", dblclick)
        .on("touchstart.zoom", touchstart);

    d3.select(window)
        .on("mousemove.zoom", mousemove)
        .on("mouseup.zoom", mouseup)
        .on("touchmove.zoom", touchmove)
        .on("touchend.zoom", touchup);
  }

  function location(point) {
    return [point[0] - x, point[1] - y, z];
  }

  // snapshot the target, data and index for subsequent dispatch
  function start() {
    target = this;
    targuments = arguments;
  }

  // Note: Since we don't rotate, it's possible for the touches to become
  // slightly detached from their original positions. Thus, we recompute the
  // touch points on touchend as well as touchstart!
  function touchup() {
    var touches = d3.svg.touches(target),
        i = -1,
        n = touches.length,
        touch;
    while (++i < n) locations[(touch = touches[i]).identifier] = location(touch);
    return touches;
  }

  function touchstart() {
    start.apply(this, arguments);
    var touches = touchup(),
        touch,
        now = Date.now();

    // doubletap detection
    if ((touches.length === 1) && (now - last < 300)) {
      var touch = touches[0];
      zoomto(1 + Math.floor(z), touch, locations[touch.identifier]);
    }
    last = now;
  }

  function touchmove() {
    var touches = d3.svg.touches(target);
    switch (touches.length) {

      // single-touch pan
      case 1: {
        var touch = touches[0];
        zoomto(z, touch, locations[touch.identifier]);
        break;
      }

      // double-touch pan + zoom
      case 2: {
        var p0 = touches[0],
            p1 = touches[1],
            p2 = [(p0[0] + p1[0]) / 2, (p0[1] + p1[1]) / 2],
            l0 = locations[p0.identifier],
            l1 = locations[p1.identifier],
            l2 = [(l0[0] + l1[0]) / 2, (l0[1] + l1[1]) / 2, l0[2]];
        zoomto(Math.log(d3.event.scale) / Math.LN2 + l0[2], p2, l2);
        break;
      }
    }
  }

  function mousedown() {
    start.apply(this, arguments);
    panning = location(d3.svg.mouse(target));
    d3.event.preventDefault();
    window.focus();
  }

  function mousemove() {
    zooming = null;
    if (panning) zoomto(z, d3.svg.mouse(target), panning);
  }

  function mouseup() {
    if (panning) {
      mousemove();
      panning = null;
    }
  }

  function mousewheel() {
    start.apply(this, arguments);

    // store starting mouse location
    if (!zooming) zooming = location(d3.svg.mouse(target));

    // detect the pixels that would be scrolled by this wheel event
    var e = d3.event, delta;
    try {
      div.scrollTop = 1000;
      div.dispatchEvent(e);
      delta = 1000 - div.scrollTop;
    } catch (error) {
      delta = e.wheelDelta || -e.detail;
    }

    zoomto(delta * .005 + z, d3.svg.mouse(target), zooming);
  }

  function dblclick() {
    start.apply(this, arguments);
    var mouse = d3.svg.mouse(target);
    zoomto(d3.event.shiftKey ? Math.ceil(z - 1) : Math.floor(z + 1), mouse, location(mouse));
  }

  function zoomto(z1, x0, x1) {
    var k = Math.pow(2, (z = z1) - x1[2]);
    x = x0[0] - k * x1[0];
    y = x0[1] - k * x1[1];
    dispatch();
  }

  function dispatch() {
    var o = d3.event, // Events can be reentrant (e.g., focus).
        k = Math.pow(2, z);

    d3.event = {
      scale: k,
      translate: [x, y],
      transform: function(sx, sy) {
        if (sx) transform(sx, x);
        if (sy) transform(sy, y);
      }
    };

    function transform(scale, o) {
      var domain = scale.__domain || (scale.__domain = scale.domain()),
          range = scale.range().map(function(v) { return (v - o) / k; });
      scale.domain(domain).domain(range.map(scale.invert));
    }

    try {
      event.zoom.dispatch.apply(target, targuments);
    } finally {
      d3.event = o;
    }

    o.preventDefault();
  }

  zoom.on = function(type, listener) {
    event[type].add(listener);
    return zoom;
  };

  return zoom;
};
})();
