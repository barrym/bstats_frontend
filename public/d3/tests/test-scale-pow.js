require("./../lib/env-js/envjs/node");
require("./../d3");

var f = d3.format(" .3f");

var x = d3.scale.pow().exponent(2);
console.log("domain([0, 1]).range([0, 1]):");
console.log("         -0.5  -> ", f(x(-0.5)));
console.log("          0.0  -> ", f(x(0.0)));
console.log("          0.5  -> ", f(x(0.5)));
console.log("          1.0  -> ", f(x(1.0)));
console.log("          1.5  -> ", f(x(1.5)));
console.log("");

var x = d3.scale.pow().exponent(2).domain([1, 2]);
console.log("domain([1, 2]).range([0, 1]):");
console.log("          0.5  -> ", f(x(0.5)));
console.log("          1.0  -> ", f(x(1.0)));
console.log("          1.5  -> ", f(x(1.5)));
console.log("          2.0  -> ", f(x(2.0)));
console.log("          2.5  -> ", f(x(2.5)));
console.log("");

var x = d3.scale.pow().exponent(2).domain([new Date(1990, 0, 1), new Date(1991, 0, 1)]);
console.log("domain([01/01/1990, 01/01/1991]).range([0, 1]):");
console.log("   10/20/1989  -> ", f(x(new Date(1989, 09, 20))));
console.log("   01/01/1990  -> ", f(x(new Date(1990, 00, 01))));
console.log("   03/15/1990  -> ", f(x(new Date(1990, 02, 15))));
console.log("   05/27/1990  -> ", f(x(new Date(1990, 04, 27))));
console.log("   01/01/1991  -> ", f(x(new Date(1991, 00, 01))));
console.log("   03/15/1991  -> ", f(x(new Date(1991, 02, 15))));
console.log("");

var x = d3.scale.pow().exponent(2).range(["red", "blue"]);
console.log("domain([0, 1]).range([\"red\", \"blue\"]):");
console.log("         -0.5  -> ", x(-0.5));
console.log("          0.0  -> ", x(0.0));
console.log("          0.5  -> ", x(0.5));
console.log("          1.0  -> ", x(1.0));
console.log("          1.5  -> ", x(1.5));
console.log("");

var x = d3.scale.pow().exponent(2).range(["red", "blue"]).interpolate(d3.interpolateHsl);
console.log("domain([0, 1]).range([\"red\", \"blue\"]).interpolate(hsl):");
console.log("        -0.50  -> ", x(-0.50));
console.log("         0.00  -> ", x(0.00));
console.log("         0.50  -> ", x(0.50));
console.log("        âˆš0.50  -> ", x(Math.SQRT1_2));
console.log("         1.00  -> ", x(1.00));
console.log("         2.00  -> ", x(2.00));
console.log("");

var x = d3.scale.pow().exponent(2);
console.log("ticks:");
console.log("            1  -> ", x.ticks(1).map(x.tickFormat(1)).join(", "));
console.log("            2  -> ", x.ticks(2).map(x.tickFormat(2)).join(", "));
console.log("            5  -> ", x.ticks(5).map(x.tickFormat(5)).join(", "));
console.log("           10  -> ", x.ticks(10).map(x.tickFormat(10)).join(", "));
console.log("");

var x = d3.scale.pow().exponent(2).domain([1, 0]);
console.log("descending ticks:");
console.log("            1  -> ", x.ticks(1).map(x.tickFormat(1)).join(", "));
console.log("            2  -> ", x.ticks(2).map(x.tickFormat(2)).join(", "));
console.log("            5  -> ", x.ticks(5).map(x.tickFormat(5)).join(", "));
console.log("           10  -> ", x.ticks(10).map(x.tickFormat(10)).join(", "));
console.log("");

var x = d3.scale.pow().exponent(2).domain([1, 2]).range([0, 1]).clamp(true);
console.log("domain clamping:")
console.log("        under  -> ", x(-1));
console.log("         over  -> ", x(2));
console.log("");

var x = d3.scale.pow().exponent(2);
console.log("nice():");
[
  [1.1, 10.9], [10.9, 1.1], [.7, 11.001], [123.1, 6.7], [0, .49],
  [.1, 1, 2.5, 3, 10.9]
].forEach(function(d) {
  var s = " [" + d.map(f) + " ]";
  while (s.length < 21) s += " ";
  console.log(" ", s + " -> ", x.domain(d).nice().domain().map(f).join(","));
});
console.log("");
