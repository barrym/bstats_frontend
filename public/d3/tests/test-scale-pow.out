domain([0, 1]).range([0, 1]):
         -0.5  ->   0.250
          0.0  ->   0.000
          0.5  ->   0.250
          1.0  ->   1.000
          1.5  ->   2.250

domain([1, 2]).range([0, 1]):
          0.5  ->  −0.250
          1.0  ->   0.000
          1.5  ->   0.417
          2.0  ->   1.000
          2.5  ->   1.750

domain([01/01/1990, 01/01/1991]).range([0, 1]):
   10/20/1989  ->  −0.194
   01/01/1990  ->   0.000
   03/15/1990  ->   0.196
   05/27/1990  ->   0.394
   01/01/1991  ->   1.000
   03/15/1991  ->   1.206

domain([0, 1]).range(["red", "blue"]):
         -0.5  ->  rgb(191,0,64)
          0.0  ->  rgb(255,0,0)
          0.5  ->  rgb(191,0,64)
          1.0  ->  rgb(0,0,255)
          1.5  ->  rgb(-319,0,574)

domain([0, 1]).range(["red", "blue"]).interpolate(hsl):
        -0.50  ->  #ffff00
         0.00  ->  #ff0000
         0.50  ->  #ffff00
        √0.50  ->  #00ff00
         1.00  ->  #0000ff
         2.00  ->  #0000ff

ticks:
            1  ->  0, 1
            2  ->  0.0, 0.5, 1.0
            5  ->  0.0, 0.2, 0.4, 0.6, 0.8, 1.0
           10  ->  0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0

descending ticks:
            1  ->  0, 1
            2  ->  0.0, 0.5, 1.0
            5  ->  0.0, 0.2, 0.4, 0.6, 0.8, 1.0
           10  ->  0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0

domain clamping:
        under  ->  0
         over  ->  1

nice():
   [ 1.100, 10.900 ]    ->   1.000, 11.000
   [ 10.900, 1.100 ]    ->   11.000, 1.000
   [ 0.700, 11.001 ]    ->   0.000, 12.000
   [ 123.100, 6.700 ]   ->   130.000, 0.000
   [ 0.000, 0.490 ]     ->   0.000, 0.500
   [ 0.100, 1.000, 2.500, 3.000, 10.900 ] ->   0.000, 1.000, 2.500, 3.000, 11.000

